app/code/Magento/Store/Model/Store.php

```diff
@@ -1175,9 +1178,14 @@ public function getCurrentUrl($fromStore = true)
             . '://'
             . $storeParsedUrl['host']
             . (isset($storeParsedUrl['port']) ? ':' . $storeParsedUrl['port'] : '')
-            . $storeParsedUrl['path']
-            . $requestString
-            . ($storeParsedQuery ? '?' . http_build_query($storeParsedQuery, '', '&amp;') : '');
+            . $storeParsedUrl['path'];
+
+        //avoid query params duplication
+        if (!preg_match('/___store=(.*?)&___from_store=(.*?)/', $requestString)) {
+            $currentUrl .= $requestString;
+        }
+
+        $currentUrl .= ($storeParsedQuery ? '?' . http_build_query($storeParsedQuery, '', '&amp;') : '');
```

Comment:
I think the source of the issue is concatenation with `?` while `$requestString` may contain query. To fix this I propose:
1. Split `$requestString` into path and query:
```php
$requestStringParts = explode('?', $requestString, 2);
$requestStringPath = $requestStringParts[0];
if (isset($requestStringParts[1])) {
   parse_string($requestStringParts[1], $requestStringQuery);
} else {
    $requestStringQuery = []
}
```
2. Merge original query params and store parsed query
```php
$currentUrlQueryParams = array_merge($requestStringQuery, $storeParsedQuery);
```
3. Safely concatenate all parts of url:
```php
$currentUrl = $storeParsedUrl['scheme']
    . '://'
    . $storeParsedUrl['host']
    . (isset($storeParsedUrl['port']) ? ':' . $storeParsedUrl['port'] : '')
    . $storeParsedUrl['path']
    . $requestStringPath
    . ($currentUrlQueryParams ? '?' . http_build_query($currentUrlQueryParams, '', '&amp;') : '');
```